vault_service/views/user.py

```diff
@@ -160,12 +158,31 @@ def store_data():
                 return '{}', 200 # or return 404?
             return q.user_data or '{}', 200
     elif request.method == 'POST':
-        d = json.dumps(payload)
-        if len(d) > MAX_ALLOWED_JSON_SIZE:
-            return json.dumps({'msg': 'You have exceeded the allowed storage limit, no data was saved'}), 400
-        u = User(id=user_id, user_data=d)
+        # limit both number of keys and length of value to keep db clean
+        if len(max(payload.values(), key=len)) > MAX_ALLOWED_JSON_SIZE:
+            return json.dumps({'msg': 'You have exceeded the allowed storage limit (length of values), no data was saved'}), 400
+        if len(payload.keys()) > MAX_ALLOWED_JSON_KEYS:
+            return json.dumps({'msg': 'You have exceeded the allowed storage limit (number of keys), no data was saved'}), 400
 
         with current_app.session_scope() as session:
+            try:
+                q = session.query(User).filter_by(id=user_id).one()
+            except ormexc.NoResultFound:
+                d = json.dumps(payload)
+                u = User(id=user_id, user_data=d)
+                try:
+                    session.add(u)
+                    session.commit()
+                    return d, 200
+                except exc.IntegrityError:
+                    session.rollback()
+                    q = session.query(User).filter_by(id=user_id).one()
+
+            data = json.loads(q.user_data)
```

Comment:
I think that between this point and the moment we save the updated data (line #189) is where a race condition can take place. I have investigated PostgreSQL >= 9.5 JSONB capabilities and I see it is possible to update a jsonb field with the || operator:

```
db=> select jsonb '{"f1":1,"f2":"test","f3":3}' || jsonb '{"f1":1,"f3":32}';
             ?column?
-----------------------------------
 {"f1": 1, "f2": "test", "f3": 32}
(1 row)
```

So, in our case we would need to execute something like this to avoid the race condition (since no read of the current user_data is needed):

```
UPDATE user 
SET user_data = user_data || jsonb '{"key1": "new value", "key3": "new value"}'
WHERE id = 35648;
```

I have not found a way to do this using sqlalchemy ORM, but we could maybe try to execute a raw sql statement:

```
session.execute('UPDATE user SET user_data = user_data || jsonb :new_user_data 
WHERE id = :user_id;', {'user_id': user_id, 'new_user_data': str(payload)})
```

I haven't tried that last code myself, but maybe you could play with it to see if you can make it work?